# 1장 3강: 데이터 특성에 따른 가설검정 기법 선택 — 실습문제

## 실습 목표

- 비교할 변수의 척도와 집단 관계를 확인할 수 있다.
- Shapiro-Wilk 검정으로 정규성을 확인할 수 있다.
- Levene 검정으로 등분산성을 확인할 수 있다.
- 가정 점검 결과에 따라 독립표본 t검정과 Welch t검정을 선택할 수 있다.
- 선택한 검정의 p-value를 해석하여 데이터에 근거한 결론을 작성할 수 있다.

## 실습 환경 / 데이터

- Python
- pandas
- scipy.stats
- `ames_housing.csv`

| 컬럼 | 의미 | 유형 |
|---|---|---|
| `SalePrice` | 주택 판매가격 | 연속형 |
| `KitchenQual` | 주방 품질 | 범주형·순서형 |
| `HeatingQC` | 난방 품질 | 범주형·순서형 |

> 모든 판단의 유의수준은 `α = 0.05`입니다.  
> 표본 추출에는 `random_state=42`를 사용하여 실행할 때마다 같은 결과가 나오도록 합니다.


## 실습 준비

1. pandas와 `scipy.stats`를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 전체 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [2]:
# 실습 준비 코드를 작성하세요.
import pandas as pd 
from scipy import stats

df = pd.read_csv(r'C:\Machine_Learning_worksheet\Data\ames_housing.csv', encoding = 'utf-8-sig')

---

## 필수 1. 정규성은 충족하지만 등분산성이 위반된 경우

### 문제 1-1. 주방 품질 `Gd`와 `TA` 집단의 판매가격 비교

#### 문제 설명

주방 품질이 `Gd`(Good)인 주택과 `TA`(Typical/Average)인 주택의 판매가격을 비교하려고 합니다. 각 집단에서 30개씩 표본을 추출한 뒤 정규성과 등분산성을 확인하고 적절한 t검정 방법을 선택하세요.

#### 요구사항

1. `KitchenQual == "Gd"`인 주택의 `SalePrice`에서 30개를 추출해 `group_gd`에 저장하세요.
2. `KitchenQual == "TA"`인 주택의 `SalePrice`에서 30개를 추출해 `group_ta`에 저장하세요.
3. 두 집단의 표본 수와 평균을 확인하세요.
4. 각 집단에 Shapiro-Wilk 정규성 검정을 수행하세요.
5. 두 집단에 Levene 등분산 검정을 수행하세요.
6. 각 가정의 p-value를 0.05와 비교하여 충족 여부를 판단하세요.
7. 다음 규칙에 따라 t검정 방법을 선택하세요.
   - 두 집단 모두 정규성 충족 + 등분산성 충족: 독립표본 t검정
   - 두 집단 모두 정규성 충족 + 등분산성 위반: Welch t검정
8. 선택한 검정을 실행하고 검정통계량과 p-value를 출력하세요.
9. 두 집단의 판매가격 차이가 통계적으로 유의한지 해석하세요.

#### 해석 질문

**Q1.** Shapiro-Wilk 검정의 귀무가설은 무엇인가요?  
**Q2.** Levene 검정의 귀무가설은 무엇인가요?  
**Q3.** 가정 점검 결과 어떤 t검정 방법을 선택해야 하나요?  
**Q4.** 선택한 검정의 결과에 따르면 두 집단의 판매가격 차이는 유의한가요?

#### 제출 결과

- 집단별 표본 수와 평균
- 정규성 및 등분산성 검정 결과
- t검정 방법 선택과 선택 근거
- 최종 검정통계량과 p-value
- 결과 해석
- Q1~Q4 답변


In [ ]:
# 필수 1 코드를 작성하세요.

### 필수 1 답변 작성란

- **Q1.**
- **Q2.**
- **Q3.**
- **Q4.**

---

## 필수 2. 정규성과 등분산성이 모두 충족된 경우

### 문제 2-1. 난방 품질 `TA`와 `Fa` 집단의 판매가격 비교

#### 문제 설명

난방 품질이 `TA`(Typical/Average)인 주택과 `Fa`(Fair)인 주택의 판매가격을 비교하려고 합니다. 각 집단에서 20개씩 표본을 추출한 뒤 정규성과 등분산성을 확인하고 적절한 t검정 방법을 선택하세요.

#### 요구사항

1. `HeatingQC == "TA"`와 `HeatingQC == "Fa"`인 집단에서 `SalePrice`를 20개씩 추출하세요.
2. 두 집단의 표본 수와 평균을 확인하세요.
3. 두 집단이 독립집단인지 대응집단인지 판단하세요.
4. 각 집단에 Shapiro-Wilk 정규성 검정을 수행하세요.
5. 두 집단에 Levene 등분산 검정을 수행하세요.
6. 두 집단 모두 정규성을 충족하는지 확인하세요.
7. 등분산성 결과에 따라 독립표본 t검정 또는 Welch t검정 중 적절한 방법을 선택하세요.
8. 선택한 검정을 실행하고 검정통계량과 p-value를 출력하세요.
9. 두 집단의 판매가격 차이가 통계적으로 유의한지 해석하세요.

#### 해석 질문

**Q1.** 두 집단은 독립집단인가요, 대응집단인가요?  
**Q2.** 두 집단의 정규성 가정은 충족되나요?  
**Q3.** 두 집단의 등분산성 가정은 충족되나요?  
**Q4.** 가정 점검 결과 어떤 t검정 방법을 선택해야 하나요?  
**Q5.** 최종 검정 결과는 무엇을 의미하나요?

#### 제출 결과

- 집단별 표본 수와 평균
- 집단 관계 판단
- 정규성 및 등분산성 검정 결과
- t검정 방법과 선택 근거
- 검정통계량과 p-value
- 결과 해석
- Q1~Q5 답변


In [ ]:
# 필수 2 코드를 작성하세요.

### 필수 2 답변 작성란

- **Q1.**
- **Q2.**
- **Q3.**
- **Q4.**
- **Q5.**

---

## 과제. 난방 품질에 따른 t검정 방법 선택

### 문제 3-1. 난방 품질 `Ex`와 `TA` 집단의 판매가격 비교

#### 문제 설명

난방 품질이 `Ex`(Excellent)인 주택과 `TA`(Typical/Average)인 주택의 판매가격을 비교하려고 합니다. 각 집단에서 20개씩 표본을 추출한 뒤 필수 문제에서 학습한 **독립표본 t검정과 Welch t검정 선택 과정**을 독립적으로 적용하세요.

#### 요구사항

1. `HeatingQC == "Ex"`인 집단과 `HeatingQC == "TA"`인 집단에서 `SalePrice`를 20개씩 추출하세요.
2. 두 집단의 표본 수와 평균을 출력하세요.
3. 두 집단이 독립집단인지 대응집단인지 판단하세요.
4. 각 집단의 정규성과 두 집단의 등분산성을 검정하세요.
5. 두 집단 모두 정규성을 충족하는지 확인하세요.
6. 등분산성 결과에 따라 독립표본 t검정 또는 Welch t검정 중 적절한 방법을 선택하고 선택 이유를 작성하세요.
7. 선택한 검정을 실행하여 검정통계량과 p-value를 출력하세요.
8. 난방 품질에 따라 판매가격에 유의한 차이가 있는지 결론을 작성하세요.

#### 해석 질문

**Q1.** 두 집단의 정규성 가정은 충족되나요?  
**Q2.** 두 집단의 등분산성 가정은 충족되나요?  
**Q3.** 최종적으로 어떤 t검정 방법을 선택해야 하나요?  
**Q4.** 검정 결과 난방 품질에 따른 판매가격 차이는 통계적으로 유의한가요?

#### 제출 결과

- 표본 구성과 집단 관계 판단
- 정규성 및 등분산성 검정 결과
- 최종 t검정 선택과 근거
- 검정통계량과 p-value
- 결과 해석
- Q1~Q4 답변


In [3]:
# 과제 코드를 작성하세요.

import numpy as np
from scipy import stats

# 1) 각 집단에서 20개씩 추출
ex = df.loc[df['HeatingQC'] == 'Ex', 'SalePrice'].dropna().sample(20, random_state=42)
ta = df.loc[df['HeatingQC'] == 'TA', 'SalePrice'].dropna().sample(20, random_state=42)

# 2) 표본 수와 평균
print(f"Ex: n = {len(ex)}, 평균 = {ex.mean():.4f}")
print(f"TA: n = {len(ta)}, 평균 = {ta.mean():.4f}")

# 4) 정규성 검정 (Shapiro-Wilk)
sw_ex = stats.shapiro(ex)
sw_ta = stats.shapiro(ta)
print(f"\n[정규성] Ex: W = {sw_ex.statistic:.4f}, p = {sw_ex.pvalue:.4f}")
print(f"[정규성] TA: W = {sw_ta.statistic:.4f}, p = {sw_ta.pvalue:.4f}")

# 4) 등분산성 검정 (Levene)
lev = stats.levene(ex, ta)
print(f"\n[등분산성] Levene: W = {lev.statistic:.4f}, p = {lev.pvalue:.4f}")

# 5~6) 판단 및 방법 선택
alpha = 0.05
norm_ok = (sw_ex.pvalue > alpha) and (sw_ta.pvalue > alpha)
eq_var  = lev.pvalue > alpha
print(f"\n정규성 충족: {norm_ok}")
print(f"등분산성 충족: {eq_var} → {'독립표본 t검정' if eq_var else 'Welch t검정'}")

# 7) 선택한 검정 실행
t_stat, p_value = stats.ttest_ind(ex, ta, equal_var=eq_var)
print(f"\n검정통계량 = {t_stat:.4f}")
print(f"p-value    = {p_value:.4f}  (지수: {p_value:.4e})")
print(f"→ {'유의한 차이 있음' if p_value <= alpha else '유의한 차이 없음'}")

Ex: n = 20, 평균 = 259451.2000
TA: n = 20, 평균 = 130845.0000

[정규성] Ex: W = 0.9447, p = 0.2939
[정규성] TA: W = 0.9736, p = 0.8290

[등분산성] Levene: W = 6.9271, p = 0.0122

정규성 충족: True
등분산성 충족: False → Welch t검정

검정통계량 = 4.9073
p-value    = 0.0000  (지수: 4.9458e-05)
→ 유의한 차이 있음


### 과제 답변 작성란

**Q1.** 두 집단의 정규성 가정은 충족되나요?  
두 집단의 정규성은 충족한다 왜냐하면 정규성은 0.05보다 높으면 되는데 두 집단의 정규성은 각각 0.2939 , 0.8290 이기 때문에

**Q2.** 두 집단의 등분산성 가정은 충족되나요?  
등분산성 가정은 False다 왜냐하면 0.05보다 작기 때문에

**Q3.** 최종적으로 어떤 t검정 방법을 선택해야 하나요?  
Welch t검정을 선택 해야한다. 왜냐하면 정규성의 결과는 가정이 충족이 되어 t 검정을 쓸 수 있지만 등분산성은 충족이 안되기 때문에 t  검정을 못쓴다. 

**Q4.** 검정 결과 난방 품질에 따른 판매가격 차이는 통계적으로 유의한가?
유의한 차이가 있다고 본다 왜냐하면 검정 통계량은 P value가 0.05보다 낮기  때문에



---

## 실습 마무리

1. 두 집단을 비교하기 전에 어떤 데이터 특성을 먼저 확인해야 하나요?
2. 정규성 검정과 등분산 검정에서 `p > 0.05`는 무엇을 의미하나요?
3. 두 집단 모두 정규성을 충족하고 등분산성도 충족하면 어떤 검정을 사용할 수 있나요?
4. 두 집단 모두 정규성을 충족하지만 등분산성이 위반되면 어떤 검정을 사용할 수 있나요?
5. 독립표본 t검정과 Welch t검정을 선택할 때 정규성과 등분산성을 함께 확인해야 하는 이유는 무엇인가요?
